Add templates for promoting features (#3)
istio/enhancements: .github/ISSUE_TEMPLATE/beta_feature.md

---
name: Beta Feature Requirements
about: Requirements for promoting a feature to beta
---

This page lists the requirements for promoting a feature to beta. Please check off and document the steps as they are completed.

**Feature Name:** 

> --- 

Reviewer: could we highlight somehow which are new to beta (from alpha)?  Seems many are redundant.

Reviewer: can we add link to demonstrate the feature has met the alpha approval?

---
name: Beta Feature Requirements
about: Requirements for promoting a feature to beta
---

This page lists the requirements for promoting a feature to beta. Please check off and document the steps as they are completed.

**Feature Name:** 
--- 

### Requirements: 

**Design**

> - [ ] Design doc has been approved

Reviewer 2: ```suggestion
- [ ] Design doc has been approved by relevant WG leads
```

**Config**

> - [ ] Can be enabled by default without requiring explicit user action. 

Reviewer 2: This isn't really a checklist item, maybe the item is to consider whether the feature should be enabled by default?

Author: This was listed in the feature requirements from istio/community. However, I'm interpreting this as saying that in alpha, features must be disabled by default. In beta, they can either be enabled or disabled by default. As such, there is a restriction on alpha, but in beta anything goes.

**Docs** 

> - [ ] Performance expectations are documented, may have caveats. 

Reviewer 2: Maybe rephrase like the following ones:

[ ] Documentation on istio.io includes performance expectations, may have caveats.

Author: Added. Thanks.

- [ ] Documentation on istio.io includes samples/tutorials. 
- [ ] Documentation on istio.io includes appropriate glossary entries. 

> - [ ] Where possible, all documentation changes have istio.io tests written. 

Reviewer 2: Drop where possible here too, and we need to make this more explicit about what is different vs. alpha.

Author: Rephrasing this as "all new documentation containing user actions include istio.io tests". The phrasing here is weird, but for instance:
* the glossary can't be tested because it only contains text
* tutorials usually have lists of actions for a user to perform and can be tested
* reference docs can sometimes be tested

- [ ] Release notes have been added. 
- [ ] Upgrade notes have been added. 

**Tests**

- [ ] Test coverage is >= 80%
- [ ] Integration tests cover feature edge cases
- [ ] End-to-end tests cover samples/tutorials
- [ ] Fixed issues have tests to prevent regressions

> - [ ] Ensure that coverage exists for the feature in stability/stress test suite.

Reviewer 2: ```suggestion
- [ ] Stability/stress test suite includes coverage for the feature.
```

**Performance**

> - [ ] Feature has baseline performance tests

Reviewer 3: what does this mean

Reviewer 4: May be:

Performance results & impacts are published with the feature enabled

**API**

> - [ ] API has had a thorough API review and is thought to be complete. 

Reviewer 3: this seems wildly subjective. Who decides if the istio/api review with 2 approvers was thorough or not, and if its "complete"?

Reviewer 3: I mean it can be subjective I guess - as long as we clarify we makes that subjective decision

**CLI**

> - [ ] Necesssary CLI command has been implemented and are complete. 

Reviewer 2: ```suggestion
- [ ] Any necessary CLI commands have been implemented and are complete. 
```

**Bugs**

> - [ ] Feature has no outstanding P0 bugs

Reviewer 3: This makes sense at first - but the reality is that even stable features have P0 bugs. Is this really fair?

**Promotion**

> - [ ] The decision to promote a feature is presented and approved by the workgroup. 

Reviewer 2: ```suggestion
- [ ] The decision to promote the feature has been presented and approved by the appropriate workgroups. 
```

> - [ ] The decision to promote a feature is approved by the TOC in the roadmap for a release. 

Reviewer 2: ```suggestion
- [ ] The decision to promote the feature has been approved by the TOC.
```

> - [ ] Supportability review addressing support scenario questions that has been discussed in the appropriate workgroup.

Reviewer 2: Need to reword this to be a bit more clear on what they need to do.

Reviewer 3: Yes please be way more specific here. I heard some discussion about a panel of 3 reviewers or something - needs a lot of clarification.

Reviewer: shouldn't the support ability review team also sign off on the promotion to beta?

---
name: Beta Feature Requirements
about: Requirements for promoting a feature to beta
---

This page lists the requirements for promoting a feature to beta. Promotion to beta must meet all requirements for promotion to alpha. Please check off and document the steps as they are completed.

**Feature Name:** 

> **Promotion to Alpha:**

Reviewer 2: Or just Alpha Checklist:

> **Promotion to Alpha:**

Author: Done

**Relevant documentation:**

```
```

--- 

### Requirements: 

**Design**

- [ ] Design doc has been approved by relevant work group leads

**Docs** 

- [ ] Documentation on istio.io includes performance expectations; may have caveats. 
- [ ] Documentation on istio.io includes samples/tutorials. 
- [ ] Documentation on istio.io includes appropriate glossary entries. 
- [ ] All new documentation containing user actions includes istio.io tests.
- [ ] Release notes have been added. 
- [ ] Upgrade notes have been added. 

**Tests**

- [ ] Integration tests cover feature edge cases
- [ ] End-to-end tests cover samples/tutorials
- [ ] Fixed issues have tests to prevent regressions
- [ ] Stability/stress test suite includes coverage for the feature.

**Performance**

> - [ ] Performance results and impacts with feature enabled are published on istio.io.

Author: I like this as a requirement for beta. I'll remove the requirement for stable since it's already present at the beta level.

> - [ ] Performance results and impacts with feature enabled are published on istio.io.

Author: This is new functionality though which we discussed in TOC, but I'm not sure we've discussed how to implement. Going to list this as:

"Tests exist with the feature enabled that can be integrated with our automated performance testing"

**API**

- [ ] TOC has reviewed the API and determined it to be complete. 

**CLI**

> - [ ] Any necessary CLI commands have been implemented and are complete. 

Reviewer 4: Tooling to debug the feature

Author: Hmmm.

Any necessary tooling to configure/use/debug/deploy feature has been implemented and is complete?

This could be isioctl commands, debug features, deployment methods, or anything else. In the community PR, it was agreed on that these would be CLI commands, but it could be anything. We might want to expand this beyond CLI as well.

---
name: Beta Feature Requirements
about: Requirements for promoting a feature to beta
---

# Beta Feature Requirements

This page lists the requirements for promoting a feature to beta. Promotion to beta must meet all requirements for promotion to alpha. Please check off and document the steps as they are completed.

**Feature Name:** 

**Design Doc:**

**Alpha Checklist:**

**Relevant Documentation:**

--- 

### Requirements: 

> **Design**

Reviewer 5: isn't this duplicate with alpha reuirement

Reviewer 5: I mean when this feature is migrated to beta without change

Author: Feature coverage and test plans aren't required for alpha. For alpha, we're just expecting an RFC with an idea of what you want to implement as well as what it it's needed for. Beta expands on that with a design doc including an expectation that you'll know how you want to implement it and have an idea of what sort of testing it might need. Some of the content overlaps between the RFC and design doc, but not all.

Author: @hzxuzhonghu does that make sense?
